### PR TITLE
Feat/pagination

### PR DESCRIPTION
**Pull request by pgeadas**
Background
=====

As our application's data grows, it becomes more difficult to deal with large datasets efficiently. We need to optimize the accesses to our database.

Problem
=====

We are encountering challenges in handling large datasets, hence we need to implement an efficient pagination mechanism that allows users to navigate through data seamlessly.

Goal
=====

Implement a pagination architecture that covers the follwoing points:

1. Pagination Interface: should support specifying the page number and the number of items per page;
2. Query Optimization: only the necessary data is fetched from the database, as fast as possible;
3. Support for Large Datasets: the solution should behave with equal performance even when there are millions of records in the database;
4. Customizable Order: users should be able to specify the order in which the data should be sorted;
5. Error Handling: Implement appropriate error handling mechanisms.

Implementation
=====

- Use a clean architecure design, maintaining a clear separation of concerns between components and allowing for easy extension and testing;
- Implement Keyset pagination for the fetch queries, that behaves equally inependently of the size of the database;
- For queries that allow it, the ones without order by, use prepared statements;
- Return appropriate error messages and codes when it is not possible to fullfil the user requests

Caveats
=====

None.

Further considerations
=====

There are some improvements to the solution that I did not have time to make but left documented, so we can tackle them in a future iteration.

I think that it might be worth mentioning that, for the purpose of this exercise, I decided to refactor the entire structure of the code and also added things that were not even asked for. If this was a real feature I would have limited myself to the scope of the issue and, if needed, creating new follow-up tasks.

### CODE REVIEW

**Comment by pgeadas** [comment]
Hi @swampie, I guess that the PR is ready for reviewing :)

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy [hunk 1]
```diff
@@ -128,7 +128,7 @@ class EventHandler extends JsonHandler {
             sendOkResponse(http, event, HttpStatus.Ok.code)
         } catch (RuntimeException ex) {
             println "Error saving event: $ex.stackTrace"
-            sendOkResponse(
+            sendErrorResponse(
```
Comment (pgeadas) [inline app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:131]: This was for the POST method which is not part of this feature.

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy [hunk 2]
```diff
@@ -64,14 +64,14 @@ class EventHandler extends JsonHandler {
...
-        if (!pageNumber) {
+        if (!pageNumber || pageNumber < 1) {
             println queryParamValidator.invalidParamsMessageOrDefault("pageNumber")
             sendErrorResponse(http, queryParamValidator.invalidParamsMessageOrDefault("pageNumber"), HttpStatus.BadRequest.code)
             return
         }
 
         Integer itemCount = queryParamValidator.validateItemCount(queryParams)
-        if (!itemCount) {
+        if (!itemCount || itemCount < 1) {
```
Comment (pgeadas) [inline app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:74]: Just a fail fast check, since this was still protected when we create the PageDetails class

File: app/src/main/groovy/io/seqera/events/domain/pagination/Pagination.groovy [hunk 3]
```diff
@@ -16,7 +16,7 @@ interface Pagination<T> {
             println "Validation Error (itemCount): $pageDetails.itemCount"
             return false
         }
-        if (pageDetails.pageNumber < 0) {
+        if (pageDetails.pageNumber <= 0) {
```
Comment (pgeadas) [inline app/src/main/groovy/io/seqera/events/domain/pagination/Pagination.groovy:19]: Same here, this will never happen since the PageDetails is protected at creation time.

File: app/src/test/groovy/io/seqera/events/application/handlers/EventHandlerIntegrationWithMockedRepoSpec.groovy [hunk 4]
```diff
@@ -112,21 +79,24 @@ class EventHandlerIntegrationWithMockedRepoSpec extends Specification {
 
         where:
         pageDetails | ordering                            | orderBy             | sort            | statusCode
-        of(1, 3)    | of(["id"], [false])                 | "orderBy=id"        | "sort=desc"     | HttpStatus.Ok.code
-        of(1, 3)    | of(["userId"], [true])              | "orderBy=userId"    | "sort=asc"      | HttpStatus.Ok.code
+        of(1, 3)    | []                                  | "orderBy=id"        | " "             | HttpStatus.Ok.code
```
Comment (pgeadas) [inline app/src/test/groovy/io/seqera/events/application/handlers/EventHandlerIntegrationWithMockedRepoSpec.groovy:82]: Merged these 3 tests in one single one since they were all testing the same

File: IMPLEMENTATION_DETAILS.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# Implementation details
+
+The solution presented makes use of two different techniques for query optimization:
+
+1. Keyset pagination
+
+   Since we have a unique identifier for each row of the database (the id), we can improve our query time (especially for large datasets) by filtering our results directly by it. This operation is usually faster than relying on the OFFSET, since the latter needs to skip N amount of rows until it reaches the desired one, which still involves some processing. This can obviously be a bigger issue for very large datasets, therefore it was avoided.
+    Additionally, since the id is the primary key which is indexed by default, using keyset pagination will be very efficient (it works better on indexed columns). Also, another advantage of using the id (which is auto incremented) is that we will not suffer from potential race condition issues like we could when we use OFFSET (since a value could be added in, between the time that we query and the time that the database seeks the desired position).
```
Comment (swampie) [inline IMPLEMENTATION_DETAILS.md:8]: Thanks for the explanation of the selected strategy for pagination. Do you think that there are any downsides of using this approch?

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy (hunk 2 above)
Comment (swampie) [reply app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:74]: 👍

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+    private static String getQueryString(HttpExchange http) {
+        return http.requestURI.query
+    }
+
+    private static String getRequestBody(HttpExchange http) {
+        return http.requestBody.text
+    }
+
+    @CompileStatic
+    class QueryParamValidator {
```
Comment (swampie) [inline app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:148]: This class could have been extracted in its own file in order to facilitate testing

File: app/src/main/groovy/io/seqera/events/domain/event/Event.groovy [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package io.seqera.events.domain.event
+
+import groovy.transform.CompileStatic
+
+@CompileStatic
+class Event {
+
+    static final def VALID_FIELD_NAMES = ["id", "workspaceId", "userId", "cpu", "mem", "io"]
```
Comment (swampie) [inline app/src/main/groovy/io/seqera/events/domain/event/Event.groovy:8]: Is usually better to add the type explicitly for class attributes instead of using only def keyword

File: app/src/main/groovy/io/seqera/events/domain/pagination/PageDetails.groovy [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    static validate(long pageNumber, int itemCount) {
+        if (pageNumber < 1) {
+            throw new IllegalArgumentException("Invalid pageNumber: ${pageNumber}")
+        }
+        if (itemCount < 1) {
+            throw new IllegalArgumentException("Invalid itemCount: ${itemCount}")
+        }
+    }
+
+    boolean equals(o) {
```
Comment (swampie) [inline app/src/main/groovy/io/seqera/events/domain/pagination/PageDetails.groovy:33]: There is a @EqualsAndHashCode annotation in groovy that would have make the code less verbose

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+        private static String validateSortingOrder(String orderBy) {
+            return orderBy == null || orderBy.isEmpty() ? "asc"
+                    : orderBy == 'asc' ? 'asc'
+                    : orderBy == 'desc' ? 'desc'
+                    : null
+        }
+
+        private static Long validatePageNumber(Map<String, String> queryParams) {
+            try {
+                return queryParams["pagenumber"] as Long
```
Comment (swampie) [inline app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:216]: Would have been beneficial for the readers to have a different casing for the parsed argument name, also there are some string repetitions in this class that could have been converted in constants

File: app/src/test/groovy/io/seqera/events/utils/QueryParamParserTest.groovy [hunk 10]
```diff
@@ -0,0 +1,72 @@
+package io.seqera.events.utils
+
+
+import spock.lang.Specification
+
+class QueryParamParserTest extends Specification {
```
Comment (swampie) [inline app/src/test/groovy/io/seqera/events/utils/QueryParamParserTest.groovy:6]: Can you elaborate why you mixed different testing engines (spock and junit)?

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+            default -> http.sendResponseHeaders(HttpStatus.MethodNotAllowed.code, 0)
+        }
+    }
+
+    void handleGet(HttpExchange http) {
+        def query = getQueryString(http)
+        println "query: $query"
+        if (!query) {
+            println queryParamValidator.missingParamsMessageOrDefault("pageNumber/itemCount")
+            sendErrorResponse(http, queryParamValidator.missingParamsMessageOrDefault("pageNumber/itemCount"), HttpStatus.BadRequest.code)
```
Comment (swampie) [inline app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:54]: This pattern (print + sendError) is repeated throughout the code: can you encapsulate the logic in a common method?

**Review by swampie — COMMENTED** [review]
Thanks for the submission: surprising groovy skills for being your first time :) Also good explanation of the changes considering the refactor that you put in place.

Code and tests were running correctly and also edge case testing where correctly handled.
We left a couple of comments in the PR for you to have a look and comment accordingly

File: IMPLEMENTATION_DETAILS.md (hunk 5 above)
Comment (pgeadas) [reply IMPLEMENTATION_DETAILS.md:8]: In the end the decision was taken after weighting all the advantages and disadvantages of both approaches, keyset and offset pagination (taking into account the size of the dataset and the probability of some of them happening).
Even though I decided to move forward with keyset pagination, there are some points where it might struggle as well like:
1. Table schema changes: If we need to change the schema of the database and decide to remove or change the "id" type or use another field as primary key, we will need to adapt our solution which involves changing the code (offset pagination does not suffer from this, as it does not care about columns) and possibly adding indexes to the columns used as keys, in order to keep our solution performant. Depending on the new "id" type, it might not even be possible to use keyset pagination, which works on the premise of having one or more columns that provide a unique and ordered identifier.
2. Missing records: We can still see some pages with less items than intended if a delete happens in the middle of a query. If we are unlucky, the delete is for a row that is already added to the ResultSet while the query is still fetching the remaining rows, so either we are fetching a value that is not there anymore or we will have one less item in the end. This depends on the strategy used for concurrency and if we are using locks or not, but considering the current implementation this could be an issue since we are not using any kind of concurrency control. (Both approaches can suffer from this, even though, for large datasets, it is more likely to happen using OFFSET since we need to skip a huge amount of rows to reach our target).
3. Sorting: We are only allowing sorting AFTER we fetch the respective number of items. Another option would be to sort BEFORE we fetch the LIMIT amount of items, but to do that efficiently we would need to have indexes in those columns. (Both approaches suffer from the lack of indexes in this case). However, I think it is worth saying that if we had more indexes, adapting the solution would be trivial: we could ask the user for an extra parameter, indicating if the sort should be done before or after the filtering. 
4. ID size: For very large datasets we would need to use a BigInt for the id, which consumes more space, but would avoid overflow which would break our solution (Offset pagination does not suffer from this)
5. Multiple-table queries: If we had multiple tables and complex join operations, maintaining the same order of results could be tricky (however, both approaches suffer from this).

These are the main disadvantages that I can think of, but even then some are shared between both approaches. Let me know if I missed some please :)

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy (hunk 6 above)
Comment (pgeadas) [reply app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:148]: I honestly thought about it, but then I also think that it can be implicitly tested as a black box while testing our Controller. This way we remove the need to have the class "separated" from the only place where it is used and we avoid having to write and maintain tests explicitly for it. However, if preferred by the team, we could extract it and test it individually or if it becomes more complex.

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy (hunk 9 above)
Comment (pgeadas) [reply app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy:216]: Yes, I agree. I will also extract those as constants then.

File: app/src/main/groovy/io/seqera/events/domain/event/Event.groovy (hunk 7 above)
Comment (pgeadas) [reply app/src/main/groovy/io/seqera/events/domain/event/Event.groovy:8]: Ok, nice to know! I thought that because this is static, the type inference would be done only once when the class is loaded.

File: app/src/test/groovy/io/seqera/events/utils/QueryParamParserTest.groovy (hunk 10 above)
Comment (pgeadas) [reply app/src/test/groovy/io/seqera/events/utils/QueryParamParserTest.groovy:6]: Sure! Brace yourself, wall text is incoming...

I started by using Mockito because it is more widely used, at least within the Java community. However, many tests were more about data and not as much about interactions between components. Since there were a lot of dynamic parameters, I had a choice: either copy and paste to create very repetitive tests or use Mockito's "@MethodSource" or "@ArgumentSource" for ParameterizedTests. I went for the latter approach, but it became very hard to read with so many parameters.

It was then that I thought about Spock: being written for Groovy mainly and providing a clean and structured way of describing what the test is doing (given-when-then and where), which is a significant improvement compared to Mockito in terms of cleanliness. Also, even though I was aware of Spock, I had not tried it before, so for me, it was the perfect opportunity to learn something new.
In the end, I ended up refactoring some tests that I had written using Mockito, but I was amazed at how clean they looked in Spock and kept using it for pretty much all tests from there on.

However, I could not get rid of Mockito completely since I found an issue with Spock that Mockito does not have: stubbing and mocking using the same object is not allowed (unless it is done in the same line, not like in Mockito where we can stub and then verifying calls on that object in the end). Even then, after trying several ways that I found online and in the documentation, I was always getting no interactions with my mock/stub class, so I had to give up. It looks like in Spock, if the class is annotated with "TestInstance.Lifecycle.PER_CLASS" and the subject is shared (see EventHandlerIntegrationWithMockedRepoSpec line 78) it does not work.

TL;DR: 
1. I never used Spock before
2. I found Spock more clean and adequate for the majority of the tests
3. Could not get rid of Mockito for tests that involved stubbing and mocking in the same shared object per class

File: app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy (hunk 11 above)
Comment (pgeadas) [reply app/src/main/groovy/io/seqera/events/application/handlers/EventHandler.groovy]: Ended up using the closure but was kind of overkill, I only needed a normal method actually :)
